ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -544,4 +530,127 @@ private boolean containsName(String name) {
       return Collections.unmodifiableCollection(idToAccountMap.values());
     }
   }
+
+  /**
+   * A {@link DataUpdater} to be used for updating {@link #FULL_ACCOUNT_METADATA_PATH} inside of
+   * {@link #updateAccounts(Collection)}
+   */
+  private class ZkUpdater implements DataUpdater<ZNRecord> {
+    private final Collection<Account> updatedAccounts;
+    private AccountInfoMap remoteAccountInfoMap;
+
+    /**
+     * @param updatedAccounts The {@link Account}s to update.
+     */
+    ZkUpdater(Collection<Account> updatedAccounts) {
+      this.updatedAccounts = updatedAccounts;
+    }
+
+    @Override
+    public ZNRecord update(ZNRecord currentData) {
+      ZNRecord newRecord;
+      if (currentData == null) {
+        logger.debug(
+            "ZNRecord does not exist on path={} in HelixPropertyStore when updating accounts. Creating a new ZNRecord.",
+            FULL_ACCOUNT_METADATA_PATH);
+        newRecord = new ZNRecord(ZN_RECORD_ID);
+      } else {
+        newRecord = currentData;
+      }
+      Map<String, String> accountMap = newRecord.getMapField(ACCOUNT_METADATA_MAP_KEY);
+      if (accountMap == null) {
+        logger.debug("AccountMap does not exist in ZNRecord when updating accounts. Creating a new accountMap");
+        accountMap = new HashMap<>();
+      }
+      try {
+        remoteAccountInfoMap = new AccountInfoMap(accountMap);
+      } catch (JSONException e) {
+        // Do not depend on Helix to log, so log the error message here.
+        logger.error("Exception occurred when building AccountInfoMap from accountMap={}", accountMap, e);
+        accountServiceMetrics.remoteDataCorruptionErrorCount.inc();
+        throw new IllegalStateException("Exception occurred when building AccountInfoMap from accountMap", e);
+      }
+      // if there is any conflict with the existing record, fail the update. Exception thrown in this updater will
+      // be caught by Helix and helixStore#update will return false.
+      if (hasConflictingAccount(updatedAccounts, remoteAccountInfoMap)) {
+        // Throw exception, so that helixStore can capture and terminate the update operation
+        throw new IllegalArgumentException(
+            "Updating accounts failed because one account to update conflicts with existing updatedAccounts");
+      } else {
+        for (Account account : updatedAccounts) {
+          try {
+            accountMap.put(String.valueOf(account.getId()), account.toJson(true).toString());
```

Comment:
except this line, nothing else has changed from the previous code. Correct?

Reply:
Yes, only this line, and how we save `remoteAccountInfoMap` to a member variable for persistence.